From acd8c8da64fe148d82942a5231f438940d090d20 Mon Sep 17 00:00:00 2001
From: Morteza Jamshidi <morteza.jamshidi@qt.io>
Date: Fri, 26 Sep 2025 12:20:57 +0200
Subject: [PATCH 154/239] MaterialStyle: Fix early out conditions for
 background and foreground

This change fixes setting the background or foreground explicitly to the
theme default. The value did not persist.

Example case for background:

The default m_hasBackground is false and background color is
globalBackground that is 0xFFFAFAFA, so if user sets the background
color to 0xFFFAFAFA the logic skips it by doing early out and doesn't
set m_hasBackground to true.

Fixes: QTBUG-140068
Change-Id: I26f1182d808354f34800dc62760e5498cbd77b3b
Pick-to: 6.8
Reviewed-by: Mitch Curtis <mitch.curtis@qt.io>
(cherry picked from commit 577682cff80796d1b3033fae3d04ea3a8dfe0e0b)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 .../material/qquickmaterialstyle.cpp          | 13 ++++----
 .../qquickmaterialstyle/data/tst_material.qml | 32 +++++++++++++++++++
 2 files changed, 38 insertions(+), 7 deletions(-)

diff --git a/src/quickcontrols/material/qquickmaterialstyle.cpp b/src/quickcontrols/material/qquickmaterialstyle.cpp
index 118ce4ac34..b88694aa7e 100644
--- a/src/quickcontrols/material/qquickmaterialstyle.cpp
+++ b/src/quickcontrols/material/qquickmaterialstyle.cpp
@@ -710,11 +710,11 @@ void QQuickMaterialStyle::setForeground(const QVariant &var)
         }
     }
 
-    m_hasForeground = true;
     m_explicitForeground = true;
-    if (m_foreground == foreground)
+    if (m_hasForeground && m_foreground == foreground)
         return;
 
+    m_hasForeground = true;
     m_customForeground = custom;
     m_foreground = foreground;
     propagateForeground();
@@ -723,7 +723,7 @@ void QQuickMaterialStyle::setForeground(const QVariant &var)
 
 void QQuickMaterialStyle::inheritForeground(uint foreground, bool custom, bool has)
 {
-    if (m_explicitForeground || m_foreground == foreground)
+    if (m_explicitForeground || (m_hasForeground == has && m_foreground == foreground))
         return;
 
     m_hasForeground = has;
@@ -791,11 +791,10 @@ void QQuickMaterialStyle::setBackground(const QVariant &var)
         custom = false;
     }
 
-    m_hasBackground = true;
     m_explicitBackground = true;
-    if (m_background == background)
+    if (m_hasBackground && m_background == background)
         return;
-
+    m_hasBackground = true;
     m_customBackground = custom;
     m_background = background;
     propagateBackground();
@@ -808,7 +807,7 @@ void QQuickMaterialStyle::setBackground(const QVariant &var)
 
 void QQuickMaterialStyle::inheritBackground(uint background, bool custom, bool has)
 {
-    if (m_explicitBackground || m_background == background)
+    if (m_explicitBackground || (m_hasBackground == has && m_background == background))
         return;
 
     m_hasBackground = has;
diff --git a/tests/auto/quickcontrols/qquickmaterialstyle/data/tst_material.qml b/tests/auto/quickcontrols/qquickmaterialstyle/data/tst_material.qml
index 56bf450111..6a21110057 100644
--- a/tests/auto/quickcontrols/qquickmaterialstyle/data/tst_material.qml
+++ b/tests/auto/quickcontrols/qquickmaterialstyle/data/tst_material.qml
@@ -288,6 +288,38 @@ TestCase {
         compare(popupObject.label2.color.toString(), popupObject.Material.textSelectionColor.toString())
     }
 
+    function test_setBackground_toGlobalDefault_propagates() {
+        let parent = createTemporaryObject(buttonComponent, testCase)
+        verify(parent)
+        let child = buttonComponent.createObject(parent)
+        verify(child)
+
+        parent.Material.theme = Material.Dark
+
+        // Explicitly set to the global default value
+        parent.Material.background = "#fafafa"
+
+        // Explicitly set background is inherited to children
+        compare(parent.Material.background, "#fafafa")
+        compare(child.Material.background, "#fafafa")
+    }
+
+    function test_setForeground_toGlobalDefault_propagates() {
+        let parent = createTemporaryObject(buttonComponent, testCase)
+        verify(parent)
+        let child = buttonComponent.createObject(parent)
+        verify(child)
+
+        parent.Material.theme = Material.Dark
+
+        // Explicitly set to the global default value
+        parent.Material.foreground = "#dd000000"
+
+        // Explicitly set foreground is inherited to children
+        compare(parent.Material.foreground, "#dd000000")
+        compare(child.Material.foreground, "#dd000000")
+    }
+
     component StyledChildWindow: Window {
         objectName: "styledChildWindow"
 
-- 
2.51.2

